### PR TITLE
[SPMD] Patch nn.Linear

### PR DESCRIPTION
**Pull request by alanwaketan**
Summary:
This pull request introduces a patched version of `torch.nn.functional.linear` that uses einsum instead of torch.matmul which will flatten the tensors to 2D and collide the sharded dimensions. The torch.matmul default behavior makes it very hard for XLA compiler to propagate the sharding annotation.

Test Plan:
PJRT_DEVICE=CPU python test/test_operations.py -v -k test_patched_linear

### CODE REVIEW

**Comment by JackCaoG** [comment]
I thought we decided to let XLA compiler handle the reshape?

File: torch_xla/distributed/fsdp/utils.py [hunk 1]
```diff
@@ -122,7 +122,9 @@ def _xla_patched_nn_linear_forward(m, input):
   return XLAPatchedLinear.apply(input, m.weight, m.bias)
 
 
-def apply_xla_patch_to_nn_linear(module):
+def apply_xla_patch_to_nn_linear(module,
```
Comment (JackCaoG) [inline torch_xla/distributed/fsdp/utils.py:125]: will this function always be called? or only when fsdp being used.

Comment (alanwaketan) [reply torch_xla/distributed/fsdp/utils.py:125]: It's called in the FSDP wrapper. For GSPMD, we don't have a wrapper so we have to call it explicitly.

File: test/test_operations.py [hunk 2]
```diff
@@ -1656,6 +1658,124 @@ def test_conv2d_backward(self):
...
 
+  def test_patched_linear_3D(self):
+    linear_cpu = nn.Linear(2, 4, bias=False)
+    input_cpu = torch.randn(4, 3, 2, requires_grad=True)
+    input_cpu.retain_grad()
+    output_cpu = linear_cpu(input_cpu)
+
+    # It looks like nn.Module.to is in-place.
+    linear = copy.deepcopy(linear_cpu).to('xla')
+    apply_xla_patch_to_nn_linear(linear, xs.xla_patched_nn_linear_forward)
```
Comment (JackCaoG) [inline test/test_operations.py:1669]: I assume if you call it once, the whole file will have the linear patched, no?

Comment (JackCaoG) [reply test/test_operations.py:1669]: nvm, I see what's happening here.

File: torch_xla/experimental/xla_sharding.py [hunk 3]
```diff
@@ -521,3 +521,42 @@ def apply(self, t: torch.Tensor):
...
+  dimensions. The torch.matmul default behavior makes it very hard for XLA compiler
+  to propagate the sharding annotation.
+
+  TODO (alanwaketan): Let's patch it on the dispatcher level.
+  """
+
+  @staticmethod
+  def forward(ctx, input, weight, bias=None):
+    # bias is an optional argument
+    ctx.save_for_backward(input, weight, bias)
```
Comment (JackCaoG) [inline torch_xla/experimental/xla_sharding.py:539]: hmm.. are we doing explicit gradient checkpointing here?

Comment (JackCaoG) [reply torch_xla/experimental/xla_sharding.py:539]: not family with how nn moudle is being written but I mainly ask because the `torch.no_grad` below.

**Review by JackCaoG — APPROVED** [review]
lgtm, thanks

File: torch_xla/experimental/xla_sharding.py (hunk 3 above)
Comment (alanwaketan) [reply torch_xla/experimental/xla_sharding.py:539]: No, it's just a convention in autograd function to save the activations. So not the gradient checkpointing or re-materialization.

**Comment by alanwaketan** [comment]
Thanks, Jack!

File: torch_xla/distributed/fsdp/utils.py (hunk 1 above)
Comment (yeounoh) [reply torch_xla/distributed/fsdp/utils.py:125]: @alanwaketan did it, using einsum instead of matmul, improve the benchmark training result? I recall that it didn't actually have an impact (or lowered MFU in @jonb377 's result?)? If so, we should make a comment explaining that this may reduce the performance, too. And +1 to not enabling by default...

**Comment by yeounoh** [comment]
> > I thought we decided to let XLA compiler handle the reshape?
> 
> For XLA compiler, I guess that's a workaround. Ultimately, I think it's right for us to handle this ourselves.

If this change is not what we used... not sure if we should merge & enable for FSDP. Iirc, didn't @AlexWertheim say that using einsum resulted in lower MFU in FSDP? @jonb377 had a similar result for GSPMD lately, no? Maybe we leave a comment so people can use this with caution on demand?

File: torch_xla/distributed/fsdp/utils.py (hunk 1 above)
Comment (jonb377) [reply torch_xla/distributed/fsdp/utils.py:125]: The lower MFU was from user-mode FSDP, I don't have a head-to-head benchmark on SPMD. @alanwaketan had some data to suggest it improves performance for SPMD.

Comment (alanwaketan) [reply torch_xla/distributed/fsdp/utils.py:125]: This is the solution in our end to stop the reshape on nn.Linear's matmul.

**Comment by alanwaketan** [comment]
> > > I thought we decided to let XLA compiler handle the reshape?
> > 
> > 
> > For XLA compiler, I guess that's a workaround. Ultimately, I think it's right for us to handle this ourselves.
> 
> If this change is not what we used... not sure if we should merge & enable for FSDP. Iirc, didn't @AlexWertheim say that using einsum resulted in lower MFU in FSDP? @jonb377 had a similar result for GSPMD lately, no? Maybe we leave a comment so people can use this with caution on demand?

I used this change for all the latest results. I don't think it introduces any performance regression on my side.
